File: jsFull/scripts/grade_only.mjs
#!/usr/bin/env node
// scripts/grade_only.mjs
// ────────────────────────────────────────────────────────────────────────────
// Adds sResult / oResult to every game in history.json WITHOUT re-analyzing.
// Picks, projections, weights, Kalman — all untouched. Just grading.
//
// Usage:
//   node scripts/grade_only.mjs              # grade + save
//   node scripts/grade_only.mjs --dry-run    # preview only
// ────────────────────────────────────────────────────────────────────────────

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const HISTORY_PATH = path.join(__dirname, "..", "data", "history.json");
const SAVE = !process.argv.includes("--dry-run");

function gradeSpread(g) {
  if (!g.sPick || g.sPick === "PASS") return null;
  if (!Number.isFinite(g.homeScore) || !Number.isFinite(g.awayScore)) return null;

  const m = g.sPick.match(/(.+?)\s+([+-])(\d+(?:\.\d+)?)/);
  if (!m) return null;
  const team = m[1].trim();
  const spread = m[2] === "-" ? -parseFloat(m[3]) : parseFloat(m[3]);
  const chosenIsHome = team === g.home;
  const margin = chosenIsHome
    ? g.homeScore - g.awayScore
    : g.awayScore - g.homeScore;
  const cover = margin + spread;
  return cover > 0 ? "WIN" : cover < 0 ? "LOSS" : "PUSH";
}

function gradeTotal(g) {
  if (!g.oPick || g.oPick === "PASS") return null;
  if (!Number.isFinite(g.homeScore) || !Number.isFinite(g.awayScore)) return null;
  if (!Number.isFinite(g.total)) return null;

  const actual = g.homeScore + g.awayScore;
  if (g.oPick === "OVER")  return actual > g.total ? "WIN" : actual < g.total ? "LOSS" : "PUSH";
  if (g.oPick === "UNDER") return actual < g.total ? "WIN" : actual > g.total ? "LOSS" : "PUSH";
  return null;
}

const store = JSON.parse(fs.readFileSync(HISTORY_PATH, "utf8"));
const runs = store.runs || [];

let graded = 0, alreadyGraded = 0, noScore = 0, noPick = 0;
let sW = 0, sL = 0, oW = 0, oL = 0;

for (const run of runs) {
  for (const g of run.games || []) {
    if (!g.sPick || g.sPick === "PASS") { noPick++; continue; }
    if (!Number.isFinite(g.homeScore) || !Number.isFinite(g.awayScore)) { noScore++; continue; }

    const sr = gradeSpread(g);
    const or = gradeTotal(g);

    if (g.sResult && g.sResult === sr) { alreadyGraded++; }
    else { graded++; }

    g.sResult = sr;
    g.oResult = or;

    if (sr === "WIN") sW++;
    if (sr === "LOSS") sL++;
    if (or === "WIN") oW++;
    if (or === "LOSS") oL++;
  }
}

const pct = (w, l) => (w + l > 0) ? (100 * w / (w + l)).toFixed(1) : "0.0";
const u = (w, l) => { const v = w * 0.9091 - l; return (v >= 0 ? "+" : "") + v.toFixed(1); };

console.log("=== Grade Only ===");
console.log("  Graded:  " + graded + " games (newly graded)");
console.log("  Already: " + alreadyGraded + " (unchanged)");
console.log("  No score:" + noScore + " (future/missing)");
console.log("  No pick: " + noPick + " (PASS/no pick)");
console.log();
console.log("  Spread: " + sW + "-" + sL + " (" + pct(sW, sL) + "%) " + u(sW, sL) + "u");
console.log("  Totals: " + oW + "-" + oL + " (" + pct(oW, oL) + "%) " + u(oW, oL) + "u");

if (SAVE) {
  const ts = new Date().toISOString().replace(/[:.]/g, "-").slice(0, 19);
  const backup = HISTORY_PATH.replace(".json", "_pre_grade_" + ts + ".json");
  fs.copyFileSync(HISTORY_PATH, backup);
  console.log("\n  Backup: " + path.basename(backup));

  fs.writeFileSync(HISTORY_PATH, JSON.stringify(store, null, 2));
  console.log("  Saved " + graded + " new grades to history.json");
} else {
  console.log("\n  DRY RUN — no changes. Remove --dry-run to save.");
}

File: pyWNBAFull/scripts/grade_only.py
#!/usr/bin/env python3
# scripts/grade_only.py
# ────────────────────────────────────────────────────────────────────────────
# Adds sResult / oResult to every game in history.json WITHOUT re-analyzing.
# Picks, projections, weights, Kalman — all untouched. Just grading.
#
# Usage:
#   python scripts/grade_only.py              # grade + save
#   python scripts/grade_only.py --dry-run    # preview only
# ────────────────────────────────────────────────────────────────────────────

import json
import math
import os
import re
import shutil
import sys
from datetime import datetime

SCRIPTS_DIR = os.path.dirname(os.path.abspath(__file__))
HISTORY_PATH = os.path.join(SCRIPTS_DIR, "..", "data", "history.json")
SAVE = "--dry-run" not in sys.argv


def grade_spread(g):
    s_pick = g.get("sPick")
    if not s_pick or s_pick == "PASS":
        return None
    home_score = g.get("homeScore")
    away_score = g.get("awayScore")
    if not isinstance(home_score, (int, float)) or not math.isfinite(home_score):
        return None
    if not isinstance(away_score, (int, float)) or not math.isfinite(away_score):
        return None

    m = re.match(r"(.+?)\s+([+-])(\d+(?:\.\d+)?)", s_pick)
    if not m:
        return None
    team = m.group(1).strip()
    spread = -float(m.group(3)) if m.group(2) == "-" else float(m.group(3))
    chosen_is_home = team == g.get("home")
    margin = (home_score - away_score) if chosen_is_home else (away_score - home_score)
    cover = margin + spread
    if cover > 0:
        return "WIN"
    elif cover < 0:
        return "LOSS"
    else:
        return "PUSH"


def grade_total(g):
    o_pick = g.get("oPick")
    if not o_pick or o_pick == "PASS":
        return None
    home_score = g.get("homeScore")
    away_score = g.get("awayScore")
    total = g.get("total")
    if not isinstance(home_score, (int, float)) or not math.isfinite(home_score):
        return None
    if not isinstance(away_score, (int, float)) or not math.isfinite(away_score):
        return None
    if not isinstance(total, (int, float)) or not math.isfinite(total):
        return None

    actual = home_score + away_score
    if o_pick == "OVER":
        return "WIN" if actual > total else ("LOSS" if actual < total else "PUSH")
    if o_pick == "UNDER":
        return "WIN" if actual < total else ("LOSS" if actual > total else "PUSH")
    return None


def main():
    with open(HISTORY_PATH, "r", encoding="utf-8") as f:
        store = json.load(f)

    runs = store.get("runs", [])

    graded = 0
    already_graded = 0
    no_score = 0
    no_pick = 0
    s_w = 0
    s_l = 0
    o_w = 0
    o_l = 0

    for run in runs:
        for g in run.get("games", []):
            s_pick = g.get("sPick")
            if not s_pick or s_pick == "PASS":
                no_pick += 1
                continue
            home_score = g.get("homeScore")
            away_score = g.get("awayScore")
            if not isinstance(home_score, (int, float)) or not math.isfinite(home_score):
                no_score += 1
                continue
            if not isinstance(away_score, (int, float)) or not math.isfinite(away_score):
                no_score += 1
                continue

            sr = grade_spread(g)
            or_ = grade_total(g)

            if g.get("sResult") and g.get("sResult") == sr:
                already_graded += 1
            else:
                graded += 1

            g["sResult"] = sr
            g["oResult"] = or_

            if sr == "WIN":
                s_w += 1
            if sr == "LOSS":
                s_l += 1
            if or_ == "WIN":
                o_w += 1
            if or_ == "LOSS":
                o_l += 1

    def pct(w, l):
        return f"{(100 * w / (w + l)):.1f}" if (w + l) > 0 else "0.0"

    def u(w, l):
        v = w * 0.9091 - l
        return f"+{v:.1f}" if v >= 0 else f"{v:.1f}"

    print("=== Grade Only ===")
    print(f"  Graded:  {graded} games (newly graded)")
    print(f"  Already: {already_graded} (unchanged)")
    print(f"  No score:{no_score} (future/missing)")
    print(f"  No pick: {no_pick} (PASS/no pick)")
    print()
    print(f"  Spread: {s_w}-{s_l} ({pct(s_w, s_l)}%) {u(s_w, s_l)}u")
    print(f"  Totals: {o_w}-{o_l} ({pct(o_w, o_l)}%) {u(o_w, o_l)}u")

    if SAVE:
        ts = datetime.now().isoformat().replace(":", "-").replace(".", "-")[:19]
        backup = HISTORY_PATH.replace(".json", f"_pre_grade_{ts}.json")
        shutil.copy2(HISTORY_PATH, backup)
        print(f"\n  Backup: {os.path.basename(backup)}")

        with open(HISTORY_PATH, "w", encoding="utf-8") as f:
            json.dump(store, f, indent=2)
        print(f"  Saved {graded} new grades to history.json")
    else:
        print("\n  DRY RUN — no changes. Remove --dry-run to save.")


if __name__ == "__main__":
    main()
